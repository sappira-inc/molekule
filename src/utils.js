import { get, kebabCase } from 'lodash';
import styled from 'styled-components';

export const themeGet = (theme, lookup, fallback) => get(theme, lookup, fallback);

export const getComponentVariant = (theme, componentName, variant) => {
  const config = themeGet(
    theme,
    `components.${componentName}.variants.${variant}`,
    theme.variants[componentName][variant]
  );
  if (!config) throw new Error(`Molekule: "${variant}" variant not found in theme...`);
  return config;
};

export const getComponentStyle = componentName => ({ theme }) => themeGet(theme, `components.${componentName}.style`);

export const getComponentClassName = ({ className, theme: { classPrefix }, variant }, name) =>
  `${className || ''} ${classPrefix}-${name} ${variant ? `${classPrefix}-${name}-${variant}` : ''}`.trim();

export const createComponent = ({ name, tag = 'div', as, style, props: defaultProps = () => ({}) }) => {
  const component = as ? styled(as) : styled[tag];

  return component.attrs(props => ({
    ...defaultProps(props),
    ...props,
    className: getComponentClassName(props, kebabCase(name)),
  }))`
    ${style}
    ${getComponentStyle(name)}
    ${({ styles = {} }) => styles[name] || {}}
  `;
};
