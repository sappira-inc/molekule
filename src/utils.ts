import kebabCase from 'lodash/kebabCase';
import get from 'lodash/get';
import styled, { CSSProp } from 'styled-components';

export const themeGet = (lookup: any, fallback?: any) => ({ theme }: any = {}) => get(theme, lookup, fallback);

export const getComponentVariant = (theme: any, componentName: string, variant: string) => {
  const config = themeGet(`variants.${componentName}.${variant}`)({
    theme,
  });

  if (!config) throw new Error(`Molekule: "${variant}" variant not found in theme...`);
  return config;
};

export const getComponentSize = (theme: any, componentName: string, size: string) =>
  themeGet(`sizes.${componentName}.${size}`, {})({ theme });

export const getComponentStyle = (componentName: string) => themeGet(`styles.${componentName}`, {});

export const getVariantStyles = (componentName: string, variant: string) =>
  themeGet(`variants.${componentName}.${variant}.style`, {});

interface GetComponentClassNameProps {
  className: string;
  theme: {
    classPrefix: string;
  };
  variant: string;
}

const getComponentClassName = (
  { className, theme: { classPrefix }, variant }: GetComponentClassNameProps,
  name: string
) => `${className || ''} ${classPrefix}-${name} ${variant ? `${classPrefix}-${name}-${variant}` : ''}`.trim();

interface CreateComponentProps<T> {
  name: string;
  tag?: keyof JSX.IntrinsicElements | React.JSXElementConstructor<any>;
  as?: React.ComponentType<any>;
  style?: ((props: any) => CSSProp) | CSSProp;
  props?: (props: T) => any;
}

export const createComponent = <
  T extends object,
  O extends keyof JSX.IntrinsicElements | React.ComponentType<any> = 'div'
>({
  name,
  tag = 'div',
  as,
  style,
  props: getBaseProps = () => ({}),
}: CreateComponentProps<T>) => {
  const component = styled<O>((as || tag) as any);

  return component.attrs<T>((props: any) => {
    const baseProps = getBaseProps(props);
    const finalProps = {
      ...baseProps,
      ...props,
      className: baseProps.className || props.className,
    };

    return {
      ...finalProps,
      className: getComponentClassName(finalProps, kebabCase(name)),
    };
  })<T>`
  ${style}
  ${getComponentStyle(name)}
  ${(p: any) => getVariantStyles(name, p.variant)}
  ${({ styles = {} }) => styles[name] || {}}
`;
};

// eslint-disable-next-line
const focusableFilter = (node: any) =>
  node.tabIndex >= 0 && !node.hasAttribute('disabled') ? NodeFilter.FILTER_ACCEPT : NodeFilter.FILTER_SKIP;

type FilterFn = (node?: any) => any;

const createTreeWalker = (root: any, currentNode: any, filterFn: FilterFn = () => NodeFilter.FILTER_ACCEPT) => {
  const treeWalker = document.createTreeWalker(
    root,
    NodeFilter.SHOW_ELEMENT,
    {
      acceptNode: filterFn,
    },
    false
  );
  treeWalker.currentNode = currentNode;
  return treeWalker;
};

export const findPreviousFocusableElement = (root: any, currentNode: any) => {
  const treeWalker = createTreeWalker(root, currentNode, focusableFilter);

  if (!treeWalker.previousNode() || treeWalker.currentNode === root) {
    return treeWalker.lastChild();
  }

  return treeWalker.currentNode;
};

export const findNextFocusableElement = (root: any, currentNode: any) => {
  const treeWalker = createTreeWalker(root, currentNode, focusableFilter);

  if (!treeWalker.nextNode() || treeWalker.currentNode === root) {
    treeWalker.currentNode = root;
    return treeWalker.firstChild();
  }

  return treeWalker.currentNode;
};

export const getNextCursorPosition = (cursorPos: any, oldValue: any, newValue: any) => {
  const cursorDiff = newValue.length - oldValue.length;
  let nextPosition = cursorPos;

  if (cursorDiff > 1 || cursorDiff < -1) {
    nextPosition += cursorDiff < 0 ? cursorDiff + 1 : cursorDiff - 1;
  }

  return Math.max(nextPosition, 0);
};
