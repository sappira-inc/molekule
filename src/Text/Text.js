import React from 'react';
import { compose, space, color, typography } from 'styled-system';
import propTypes from '@styled-system/prop-types';
import { createComponent } from '../utils';

const StyledText = createComponent({
  name: 'Text',
  tag: 'span',
  style: compose(
    space,
    color,
    typography
  ),
});

const Text = React.forwardRef((props, ref) => <StyledText {...props} ref={ref} />);

Text.propTypes = {
  ...propTypes.typography,
  ...propTypes.color,
  ...propTypes.space,
};

export default Text;
