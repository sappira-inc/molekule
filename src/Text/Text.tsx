import { compose, space, color, typography, SpaceProps, ColorProps, TypographyProps } from 'styled-system';
import { createComponent } from '../utils';

export interface TextProps extends SpaceProps, ColorProps, TypographyProps {}

export const Text = createComponent<TextProps, 'span'>({
  name: 'Text',
  tag: 'span',
  style: compose(space, color, typography),
});
