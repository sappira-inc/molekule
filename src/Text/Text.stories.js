import React from 'react';
import Text from './Text';

export default {
  title: 'Components|Text',
  component: Text,
};

export const Basic = () => (
  <>
    <Text>The basic text component renders as a {'<span />'}</Text>
    <Text as="h1">It can render as other tags, like this {'<H1/>'}</Text>
    <Text color="red">It can use colors from the theme</Text>
  </>
);
