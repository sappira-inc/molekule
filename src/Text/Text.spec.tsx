import React from 'react';
import { renderWithTheme } from '../test/utils';
import Text from './Text';

describe('Text', () => {
  test('default to <span>', () => {
    const { asFragment } = renderWithTheme(<Text>Hey!</Text>);
    expect(asFragment()).toMatchSnapshot();
  });

  test('"as" prop', () => {
    const { asFragment } = renderWithTheme(<Text as="h1">Hey!</Text>);
    expect(asFragment()).toMatchSnapshot();
  });

  test('"color" prop', () => {
    const { asFragment } = renderWithTheme(<Text color="red">Hey!</Text>);
    expect(asFragment()).toMatchSnapshot();
  });
});
