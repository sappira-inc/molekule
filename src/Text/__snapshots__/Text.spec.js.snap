// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text "as" prop 1`] = `
<DocumentFragment>
  <h1
    class="re-text "
  >
    Hey!
  </h1>
</DocumentFragment>
`;

exports[`Text "color" prop 1`] = `
<DocumentFragment>
  .c0 {
  color: #FD575D;
}

<span
    class="re-text c0"
    color="red"
  >
    Hey!
  </span>
</DocumentFragment>
`;

exports[`Text default to <span> 1`] = `
<DocumentFragment>
  <span
    class="re-text "
  >
    Hey!
  </span>
</DocumentFragment>
`;
